<a href="https://colab.research.google.com/github/v0ltax/TT-2C2025-Data-Analitycs-Notebooks/blob/main/Robledo%20Sebastian%20-%20Comisi%C3%B3n%20252813%20-%20TPI%20Data%20Analytics/Pre-entrega.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Análisis de datos con Pyhon - Pre-entrega
### Nombre y apellido: Sebastian Hernan Ariel Robledo

---
# Etapa 1

## Importar Datasets
### Actividad 1

In [7]:
# Carga de datasets alojados en github
gitRepo = "https://raw.githubusercontent.com/v0ltax/TT-2C2025-Data-Analitycs-Notebooks"
urlClientes = gitRepo + "/refs/heads/main/Robledo%20Sebastian%20-%20Comisi%C3%B3n%20252813%20-%20TPI%20Data%20Analytics/clientes.csv"
urlMarketing = gitRepo + "/refs/heads/main/Robledo%20Sebastian%20-%20Comisi%C3%B3n%20252813%20-%20TPI%20Data%20Analytics/marketing.csv"
urlVentas = gitRepo + "/refs/heads/main/Robledo%20Sebastian%20-%20Comisi%C3%B3n%20252813%20-%20TPI%20Data%20Analytics/ventas.csv"

import pandas as pd

df_clientes = pd.read_csv(urlClientes)
df_marketing = pd.read_csv(urlMarketing)
df_ventas = pd.read_csv(urlVentas)

In [8]:
df_ventas

id_venta                producto   precio  cantidad fecha_venta  \
0          792       Cuadro decorativo   $69.94       5.0  02/01/2024   
1          811         Lámpara de mesa  $105.10       5.0  02/01/2024   
2         1156                Secadora   $97.96       3.0  02/01/2024   
3         1372                Heladera  $114.35       8.0  02/01/2024   
4         1546                Secadora  $106.21       4.0  02/01/2024   
...        ...                     ...      ...       ...         ...   
3030      1837         Horno eléctrico  $104.12       9.0  30/12/2024   
3031      2276                  Laptop   $85.27       9.0  30/12/2024   
3032      2696                  Laptop  $107.81       4.0  30/12/2024   
3033      2913              Smartphone   $99.85       7.0  30/12/2024   
3034      2930  Consola de videojuegos   $55.47       6.0  30/12/2024   

              categoria  
0            Decoración  
1            Decoración  
2     Electrodomésticos  
3     Electrodomésticos  
4     Electrodomésticos  
...                 ...  
3030  Electrodomésticos  
3031        Electrónica  
3032        Electrónica  
3033        Electrónica  
3034        Electrónica  

[3035 rows x 6 columns]

## Script básico
### Actividades 2 y 3

In [9]:
# Generamos una lista de diccionarios que contenga algunos datos
ventas = [
  {'producto': 'Laptop', 'precio': 500.0, 'cantidad': 3},
  {'producto': 'Heladera', 'precio': 1000.0, 'cantidad': 5},
  {'producto': 'Smartphone', 'precio': 100.0, 'cantidad': 15}
]

In [10]:
# Calculamos las ventas acumuladas - map
ventas_totales = sum(map(lambda x: x['precio'] * x['cantidad'], ventas))
ventas_totales

8000.0

In [11]:
# Calculamos las ventas acumuladas - listas por comprensión
venta_acumuladas = sum([venta["precio"] * venta["cantidad"] for venta in ventas])
venta_acumuladas

8000.0

## Dataframe ventas
### Actividad 4
### Análisis exploratorio de **datos**

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [12]:
# Aplicamos info() para familiarizarnos con el dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [13]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_ventas.head()

id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos

In [14]:
# Usar sample() para tomar registros al azar
df_ventas.sample(5)

id_venta         producto   precio  cantidad fecha_venta  \
314      1064       Aspiradora   $65.00       9.0  06/02/2024   
903      1435         Lavadora   $37.20      10.0  10/04/2024   
72        830  Adorno de pared   $67.23      10.0  10/01/2024   
966      2615      Auriculares   $46.72      11.0  16/04/2024   
665      1209         Batidora  $111.44       3.0  17/03/2024   

             categoria  
314  Electrodomésticos  
903  Electrodomésticos  
72          Decoración  
966        Electrónica  
665  Electrodomésticos

In [15]:
# Analizamos el campo precio
# Si es numericos se puede aplicar describe()
df_ventas["precio"].describe()

count       3033
unique      2590
top       $76.32
freq           5
Name: precio, dtype: object

In [16]:
# Analizamos el campo cantidad
# si es numerico se puede aplicar describe()
df_ventas["cantidad"].sample(5)

1717    6.0
2849    6.0
76      6.0
2985    8.0
1910    2.0
Name: cantidad, dtype: float64

In [17]:
# Analizamos el campo fecha_venta
df_ventas["fecha_venta"]

0       02/01/2024
1       02/01/2024
2       02/01/2024
3       02/01/2024
4       02/01/2024
           ...    
3030    30/12/2024
3031    30/12/2024
3032    30/12/2024
3033    30/12/2024
3034    30/12/2024
Name: fecha_venta, Length: 3035, dtype: object

In [18]:
# Analizamos el campo categoria
# Probar con value_counts y unique
df_ventas["categoria"].value_counts()

categoria
Decoración           1015
Electrónica          1012
Electrodomésticos    1008
Name: count, dtype: int64

### Actividad 5
### Calidad de datos

In [19]:
# Buscamos registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [20]:
# Visualizamos los duplicados
# keep: 'first', 'last', False
df_ventas[df_ventas.duplicated(keep=False)].sort_values(by="id_venta")

id_venta           producto  precio  cantidad fecha_venta    categoria
820        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
826        56           Cortinas  $66.24       5.0  05/04/2024   Decoración
832       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
844       145           Alfombra  $41.72       8.0  06/04/2024   Decoración
878       204  Jarrón decorativo  $94.75       1.0  08/04/2024   Decoración
..        ...                ...     ...       ...         ...          ...
830      2545        Auriculares  $32.81      11.0  05/04/2024  Electrónica
883      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
889      2705        Auriculares  $87.29       2.0  08/04/2024  Electrónica
831      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica
825      2778         SmartWatch  $70.66       7.0  05/04/2024  Electrónica

[70 rows x 6 columns]

In [21]:
# Buscamos datos nulos
df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

In [22]:
# Visualizar los nulos
# df_ventas[df_ventas.isnull().any(axis=1)]
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [23]:
# Analizar los precios de los NaN a ver que opciones tenemos
# df_ventas[df_ventas["producto"]=="Elementos de cerámica"]
df_ventas["producto"].value_counts()
# df_ventas["producto"].unique()


producto
Lámpara de mesa           181
Auriculares               147
Microondas                135
Cafetera                  118
Laptop                    104
Secadora                  103
Jarrón decorativo         102
Smartphone                102
Alfombra                  102
SmartWatch                102
Aspiradora                102
Cortinas                  102
Rincón de plantas         102
Freidora eléctrica        101
Espejo decorativo         101
Adorno de pared           101
Lavadora                  100
Consola de videojuegos    100
Cuadro decorativo         100
Horno eléctrico           100
Plancha de vapor          100
Proyector                 100
Elementos de cerámica     100
Batidora                  100
Televisor                 100
Cámara digital            100
Parlantes Bluetooth        99
Tablet                     58
Heladera                   49
Candelabro                 24
Name: count, dtype: int64

### Observaciones
* Cantidad de registros: 3035
* Columnas 6
* 35 duplicados, eliminarlos
* Algunos nulos (2), 1% del faltante, dropna
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64
* Campo fecha: ajustar a dtype datetime

## Dataframe marketing

### Analisis exploratorio de datos

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [24]:
# Aplicamos info() para familiarizarnos con el dataframe
df_marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


In [25]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_marketing.head()

id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.25   31/03/2024  05/05/2024

In [26]:
# Analizamos el campo costo
df_marketing["costo"].sample(5)

84    5.82
66    4.78
74    4.52
75    4.64
33    4.38
Name: costo, dtype: float64

In [27]:
# Analizamos el campo fecha_inicio
df_marketing["fecha_inicio"].sample(5)

2     28/03/2024
83    23/11/2024
62    24/09/2024
31    06/07/2024
61    22/09/2024
Name: fecha_inicio, dtype: object

In [28]:
# Analizamos el campo fecha_fin
df_marketing["fecha_fin"].sample(5)

0     03/05/2024
72    11/12/2024
27    15/07/2024
1     13/05/2024
55    07/10/2024
Name: fecha_fin, dtype: object

### Calidad de los datos

In [29]:
# Buscamos registros duplicados
df_marketing.duplicated().sum()

np.int64(0)

In [30]:
# Buscamos registros nulos
df_marketing.isnull().sum()

id_campanha     0
producto        0
canal           0
costo           0
fecha_inicio    0
fecha_fin       0
dtype: int64

### Observaciones
* Cantidad de registros: 90
* Columnas: 6
* 0 duplicados
* 0 nulos
* Campo fecha_inicio: ajustar a dtype datetime
* Campo fecha_fin: ajustar a dtype datetime

## Dataframe clientes

### Analisis exploratorio de datos

In [31]:
# Aplicamos info() para familiarizarnos con el dataframe
df_clientes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 22.3+ KB


In [32]:
# para ver los datos head y tail del dataframe
df_clientes

id_cliente               nombre  edad         ciudad  ingresos
0             1      Aloysia Screase    44  Mar del Plata  42294.68
1             2  Kristina Scaplehorn    25        Posadas  24735.04
2             3       Filip Castagne    50    Resistencia  35744.85
3             4          Liuka Luard    39   Bahía Blanca  27647.96
4             5        Dore Cockshtt    28        Rosario  28245.65
..          ...                  ...   ...            ...       ...
562         563        Dione Forsyde    29        Posadas  26757.73
563         564          Fleming Gow    39       Santa Fe  43674.96
564         565      Jewelle Mabbett    33        Córdoba  30522.64
565         566          Lauri Munns    23    Resistencia  31259.14
566         567          Micah Matis    31     Corrientes  42927.86

[567 rows x 5 columns]

### Calidad de los datos

In [33]:
# Buscamos registros duplicados
df_clientes.duplicated().sum()

np.int64(0)

In [34]:
# Buscamos datos nulos
df_clientes.isnull().sum()

id_cliente    0
nombre        0
edad          0
ciudad        0
ingresos      0
dtype: int64

### Observaciones
* Cantidad de registros: 567
* Columnas: 5
* 0 duplicados
* 0 nulos

---
# Etapa 2

## Limpieza de datos
Actividad 1 - Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.

### Dataframe ventas

* 35 duplicados, eliminarlos - se eliminaron 35 duplicados
* Algunos nulos (2), 1% del faltante, dropna - se eliminaron 2 registros con 2 NaN
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64 - Ok
* Campo fecha: ajustar a dtype datetime - Ok

In [35]:
# Eliminamos registros duplicados
# atributos a evaluar, keep, ignore_index, inplace
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [36]:
# Validamos que no haya duplicados
df_ventas.duplicated().sum()

np.int64(0)

In [38]:
# Eliminamos registros con datos nulos
df_ventas.dropna(inplace=True)

In [37]:
# Validamos que no haya nulos
# df_ventas.isnull().sum()
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3000 non-null   int64  
 1   producto     3000 non-null   object 
 2   precio       2998 non-null   object 
 3   cantidad     2998 non-null   float64
 4   fecha_venta  3000 non-null   object 
 5   categoria    3000 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 140.8+ KB


In [39]:
# Eliminamos el caracter especial $ de precio
# Normalización: usar replace()
# De ser necesario, podemos usar el atributo regex=True
df_ventas["precio"] = df_ventas["precio"].str.strip().str.replace("$", "")

In [40]:
# Validamos que precio no contenga $
df_ventas["precio"].sample(3)

153      76.92
1276     96.20
2677    120.80
Name: precio, dtype: object

In [41]:
# Convertimos precio de str a float
# Podemos usar pd.to_numeric o astype o incluso combinarlos
df_ventas["precio"] = df_ventas["precio"].astype(float)

In [42]:
# Convertir el dtype de cantidad, de float64 a int64
df_ventas["cantidad"]= df_ventas["cantidad"].astype(int)

In [43]:
# Validar la conversion a int
df_ventas["cantidad"].dtypes

dtype('int64')

In [44]:
# Convertir el campo fecha_venta a dtype datetime
# pd.to_datetime(df_ventas["fecha_venta"], format="%d/%m/%Y")
# pd.to_datetime(df_ventas["fecha_venta"], dayfirst=True)
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [45]:
# Validar la conversion de fecha
df_ventas['fecha_venta'].dtype

dtype('<M8[ns]')

### Dataframe marketing

*** Acciones ***

In [ ]:
# Convertir el campo fecha_inicio a dtype datetime


In [ ]:
# Convertir el campo fecha_fin a dtype datetime


In [ ]:
# Validamos los cambios


## Transformación de datos
Actividad 2 - Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.

In [ ]:
df_ventas.head(2)

In [ ]:
# Transformación:
# Calculamos el valor de cada venta = precio * cantidad
# Agregamos una nueva columnas
df_ventas["valor_venta"] = df_ventas["precio"] * df_ventas["cantidad"]

In [ ]:
# Visualizamos
df_ventas.head(2)

In [ ]:
# Agregación:
# Agrupar valor_venta por producto
# Sumamos las ventas dentro de cada grupo
tabla_ventas = (df_ventas.groupby("producto", as_index=False)["valor_venta"].sum())
tabla_ventas.head()

In [ ]:
# Usemos describe para analizar algunos valores del dataframe resultante
tabla_ventas.describe()

In [ ]:
# Agregamos un plot para visualizar graficamente la distribución
import seaborn as sns
import matplotlib.pyplot as plt

sns.histplot(tabla_ventas["valor_venta"], bins=30, kde=True, edgecolor="black")
plt.title("Distribución de Valor de Venta con Curva de Densidad")
plt.xlabel("Valor de venta")
plt.ylabel("Frecuencia")
plt.show()

In [ ]:
# Calcular el percentil 75 (umbral de alto rendimiento)
umbral_ventas = tabla_ventas["valor_venta"].quantile(0.75)
umbral_ventas

In [ ]:
# Identificar productos que superan el umbral
# productos_top = tabla_ventas.query("valor_venta >= @umbral")
productos_top = tabla_ventas[tabla_ventas["valor_venta"] >= umbral_ventas]
productos_top

## Agregación
Actividad 3 - Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [ ]:
df_ventas.columns

In [ ]:
# Agrupamos por categoria y agregamos la suma
ventas_categoria = df_ventas.groupby("categoria", as_index=False)["valor_venta"].sum()
ventas_categoria.head()

In [ ]:
ventas_categoria

In [ ]:
# Ordenar y mostrar los resultados
ventas_categoria

## Integración de datos
Actividad 4 - Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

In [ ]:
df_ventas.head(2)

In [ ]:
df_marketing.info()

In [ ]:
# Unir los datasets ventas y marketing por columna 'producto'
ventas_marketing = pd.merge(df_ventas, df_marketing, on="producto", how="left")

In [ ]:
# Verificar coincidencias (muestra)
ventas_marketing.head()

### Analizamos el impacto de la campaña

In [ ]:
# Nos quedamos solo con los registros de ventas durante la campaña de marketing
# Seleccionamos solo los campos que nos interesan: "producto", "valor_venta", "costo"


In [ ]:
# Agrupamos por producto y hacemos una agregación por "valor de venta" y "costo"
ventas_marketing.groupby("categoria")[["valor_venta","costo"]].sum()

In [ ]:
# Generamos una nueva columna con la diferencia entre lo vendido y lo invertido en la campaña
